components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIKeyValidator.java

```diff
@@ -331,45 +331,45 @@ public VerbInfoDTO findMatchingVerb(MessageContext synCtx) throws ResourceNotFou
 
         String httpMethod = (String)((Axis2MessageContext) synCtx).getAxis2MessageContext().
                 getProperty(Constants.Configuration.HTTP_METHOD);
-
-        API selectedApi = null;
-        Resource selectedResource = null;
-
-        for(API api : synCtx.getConfiguration().getAPIs()){
-            if(apiContext.equals(api.getContext()) && apiVersion.equals(api.getVersion())){
-                if(log.isDebugEnabled()){
-                    log.debug("Selected API: ".concat(apiContext).concat(", Version: ").concat(apiVersion));
+        if (resourceString == null) {
+            API selectedApi = null;
+            Resource selectedResource = null;
+
+            for (API api : synCtx.getConfiguration().getAPIs()) {
+                if (apiContext.equals(api.getContext()) && apiVersion.equals(api.getVersion())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Selected API: ".concat(apiContext).concat(", Version: ").concat(apiVersion));
```

Comment:
Use + or StringBuilder or String.format (if you append large objects) as a best practice to concat string. Improves clarity and performance.